src/form.js

```diff
@@ -274,23 +276,63 @@ define([
-                // no ref counting for instances without a `src` since they cannot be dropped
-                this.instanceMetadata.push(InstanceMetadata({
-                    src: attrs.src,
-                    id: attrs.id
-                }));
-                return attrs.id;
+            function getUniqueId(id, ids) {
+                var temp = (id || "data") + "-",
+                    num = 1;
+                while (!id || ids.hasOwnProperty(id)) {
+                    id = temp + num;
+                    num++;
+                }
+                return id;
             }
-
-            var meta = _.find(this.instanceMetadata, function (m) {
+            var meta;
+            if (attrs.src) {
+                meta = _.find(this.instanceMetadata, function (m) {
                     return m.attributes.src === attrs.src;
                 });
+                if (!meta) {
+                    // attrs.src not found, try to find by id
+                    var ids = _.indexBy(this.instanceMetadata, function (m) {
+                            return m.attributes.id;
+                        });
+                    meta = attrs.id && ids.hasOwnProperty(attrs.id) ? ids[attrs.id] : null;
+                    if (meta && meta.internal) {
+                        // assign new src to internal instance
+                        meta.internal = false;
+                        meta.attributes.src = attrs.src;
+                    } else {
+                        // assign unique id to attrs (will create new meta)
+                        attrs = _.clone(attrs);
+                        attrs.id = getUniqueId(attrs.id, ids);
+                        meta = null;
+                    }
+                    this.knownInstances[attrs.id] = attrs.src;
+                }
+            } else if (attrs.id) {
+                // attrs has no src, find by id
+                meta = _.find(this.instanceMetadata, function (m) {
+                    return m.attributes.id === attrs.id;
+                });
+                if (meta) {
+                    if (meta.internal && this.knownInstances.hasOwnProperty(attrs.id)) {
+                        meta.internal = false;
+                        meta.attributes.src = this.knownInstances[attrs.id];
+                    }
+                } else if (this.knownInstances.hasOwnProperty(attrs.id)) {
+                    attrs.src = this.knownInstances[attrs.id];
+                }
+            } else {
+                throw new Error("unsupported: non-primary instance without id or src");
+            }
             if (!meta) {
-                this.instanceMetadata.push(InstanceMetadata({
+                meta = InstanceMetadata({
                     src: attrs.src,
                     id: attrs.id
-                }, null, mug || null, property));
+                }, null, mug || null, property);
+                if (!attrs.src) {
+                    meta.internal = true;
```

Comment:
"internal" instance metadata (without a `src` attribute) is not serialized to XML